src/main/scala/viper/gobra/frontend/Desugar.scala

```diff
@@ -1467,20 +1577,99 @@ object Desugar {
                 (recv, args) <- dRecvWithArgs
                 convertedArgs = convertArgs(args)
                 mproxy = getMethodProxy(base, recv, convertedArgs)
-              } yield Right(in.PureMethodCall(recv, mproxy, convertedArgs, resT)(src))
+                spec = p.maybeSpec.map(closureSpecD(ctx, info))
+              } yield Right(pureMethodCall(recv, mproxy, convertedArgs, spec, resT))
             } else {
               for {
                 (recv, args) <- dRecvWithArgs
                 convertedArgs = convertArgs(args)
                 mproxy = getMethodProxy(base, recv, convertedArgs)
-              } yield Left((targets, in.MethodCall(targets, recv, mproxy, convertedArgs)(src)))
+                spec = p.maybeSpec.map(closureSpecD(ctx, info))
+              } yield Left((targets, methodCall(targets, recv, mproxy, convertedArgs, resT, spec)))
             }
           }
           case sym => Violation.violation(s"expected symbol with arguments and result or a built-in entity, but got $sym")
         }
       }
     }
 
+    /** Returns either a tuple with targets and call statement or, if the call is pure, the call expression directly. */
+    def closureCallDAux(ctx: FunctionContext, info: TypeInfo)(expr: PInvoke)(src: Meta): Writer[Either[(Vector[in.LocalVar], in.Stmt), in.Expr]] = {
```

Comment:
To me it looks like `functionCallDAux` is already taking care of the case that the call has a spec. What is this method doing that `functionCallDAux` is not already doing? If it is not doing anything additionally, I would remove this method to avoid code duplication.

Reply:
`closureCallDAux` is for closure calls, i.e. calls c(_) as spec where c is a closure expression (and not a function or method name). These need to be handled separately (the same way as function calls and method calls are handled separately within `functionCallDAux`). I have added a new method for this since there is not much in `functionCallDAux` that would be reused by this (except the code that I have factored out in `functionCallArgsD`)

Reply:
I have added a line to the comment to make this clearer